### PR TITLE
BUG: Skip good_enough when auto_max_age is None

### LINKED ISSUE #12708: iers.conf.auto_max_age = None produces distracting warnings
### Description

Working in an HPC environment, we routinely follow the guidelines for [using IERS code offline](https://docs.astropy.org/en/stable/utils/iers.html#working-offline), including setting `iers.conf.auto_max_age = None`.  Previously, in #10126, we fixed some numerical issues with comparing floating point numbers to `None`.  However, that did not fix some overflow warnings which are distracting, to the point where some colleagues thought that actual time calculations were being corrupted (they are not).

### Expected behavior

`astropy.utils.iers.iers.LeapSeconds.auto_open()` should be able to handle `iers.conf.auto_max_age = None` gracefully.

In other words, if `iers.conf.auto_max_age = None`, simply pick a value for `good_enough` (in `auto_open()`, see below for link) that doesn't generate an overflow warning.  I actually contributed #10126, but I can't recall why that was not handled at that time.

### Actual behavior

`iers.conf.auto_max_age = None` is converted into a numerical value, `np.finfo(float).max`, which is appropriate in other parts of the code, but not the calculation in `astropy.utils.iers.iers.LeapSeconds.auto_open()`.

### Steps to Reproduce

The snippet below reproduces the code in `astropy.utils.iers.iers.LeapSeconds`, in particular, [this line](https://github.com/astropy/astropy/blob/2984fa2ea5f3604b71df8ca687c59d34913e62f1/astropy/utils/iers/iers.py#L986):
```python
>>> import astropy
>>> astropy.__version__
'4.3.1'
>>> from astropy.time import Time, TimeDelta
>>> from datetime import datetime
>>> from astropy.utils.iers.iers import _none_to_float
>>> s = '{0.year:04d}-{0.month:02d}-{0.day:02d}'.format(datetime.utcnow())
>>> today  = Time(s, scale='tai', format='iso', out_subfmt='date')
>>> TimeDelta(180-_none_to_float(None), format='jd')
/Users/weaver/Documents/local/products/venv/astropy4/lib/python3.8/site-packages/astropy/time/utils.py:179: RuntimeWarning: overflow encountered in double_scalars
  c = 134217729. * a  # 2**27+1.
/Users/weaver/Documents/local/products/venv/astropy4/lib/python3.8/site-packages/astropy/time/utils.py:181: RuntimeWarning: invalid value encountered in double_scalars
  ah = c - abig
<TimeDelta object: scale='None' format='jd' value=nan>
>>> today + TimeDelta(180-_none_to_float(None), format='jd')
/Users/weaver/Documents/local/products/venv/astropy4/lib/python3.8/site-packages/astropy/time/utils.py:179: RuntimeWarning: overflow encountered in double_scalars
  c = 134217729. * a  # 2**27+1.
/Users/weaver/Documents/local/products/venv/astropy4/lib/python3.8/site-packages/astropy/time/utils.py:181: RuntimeWarning: invalid value encountered in double_scalars
  ah = c - abig
/Users/weaver/Documents/local/products/venv/astropy4/lib/python3.8/site-packages/erfa/core.py:16224: RuntimeWarning: invalid value encountered in d2dtf
  iy, im, id, ihmsf, c_retval = ufunc.d2dtf(scale, ndp, d1, d2)
<Time object: scale='tai' format='iso' value=-4713-11-24>
```

### System Details

```
Numpy 1.21.4
pyerfa 2.0.0.1
astropy 4.3.1
Scipy 1.7.2
Matplotlib 3.5.0
```

As far as I can tell the relevant code has not changed from 4.3.1 to 5.0.

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to lessen unnecessary warnings when `iers.conf.auto_max_age` is `None`.

Fixes #12708

### TODO

- [x] See if others think this is the correct solution.
- [x] Add test(s).
- [x] Add change log.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/utils/iers/iers.py
Comment (mhvk) [inline astropy/utils/iers/iers.py:986]: I think one would still want to get the best file rather than the first that is available so I would write
```
offset = 180 - (30 if conf.auto_max_age is None else conf.auto_max_age)
good_enough = cls._today() + TimeDelta(180-offset, format='jd')
```

File: astropy/utils/iers/iers.py
Comment (mhvk) [inline astropy/utils/iers/iers.py:1027]: Keep this as is to still pick the best file, and then at line 1034 only warn if the max age is set:
```
if conf.auto_max_age is not None and self.expires < self._today():
    warn(...)
```

**Review by mhvk — COMMENTED** [review]
Thanks for looking into this! A suggestion for a slightly different approach.  Obviously, this would also need a test...

File: astropy/utils/iers/iers.py
Comment (pllim) [reply astropy/utils/iers/iers.py]: Done.

File: astropy/utils/iers/iers.py
Comment (pllim) [reply astropy/utils/iers/iers.py]: Done. Thanks!

**Comment by pllim** [comment]
These tests failed for me locally but I cannot tell if my local data files are outdated or if it is real. I guess I'll wait for CI to tell me.

`pytest  astropy/utils/iers/tests/test_leap_second.py --remote-data`

```
self = <astropy.utils.iers.tests.test_leap_second.TestAutoOpenExplicitLists object at ...>
tmpdir = local('/tmp/.../pytest-16/test_fake_expired_file0')

    def test_fake_expired_file(self, tmpdir):
        fake_file1 = make_fake_file('28 June 2010', tmpdir)
        fake_file2 = make_fake_file('27 June 2012', tmpdir)
        # Between these and the built-in one, the built-in file is best.
        ls = iers.LeapSeconds.auto_open([fake_file1, fake_file2,
                                         iers.IERS_LEAP_SECOND_FILE])
        assert ls.meta['data_url'] == iers.IERS_LEAP_SECOND_FILE

        # But if we remove the built-in one, the least expired one will be
        # used and we get a warning that it is stale.
        with pytest.warns(iers.IERSStaleWarning):
            ls2 = iers.LeapSeconds.auto_open([fake_file1, fake_file2])
        assert ls2.meta['data_url'] == fake_file2
        assert ls2.expires == Time('2012-06-27', scale='tai')

        # Use the fake files to make sure auto_max_age is safe.
        # Should have no warning here.
        with iers.conf.set_temp('auto_max_age', None):
            ls3 = iers.LeapSeconds.auto_open([fake_file1,
                                              iers.IERS_LEAP_SECOND_FILE])
>       assert ls3.meta['data_url'] == fake_file1
E       AssertionError: assert '/home/.../Do...ap_Second.dat' == '/tmp/pytest-...p_seconds.dat'
E         - /tmp/.../pytest-16/test_fake_expired_file0/fake_leap_seconds.dat
E         + /home/.../astropy/utils/iers/data/Leap_Second.dat

astropy/utils/iers/tests/test_leap_second.py:193: AssertionError
_________________________________________ TestDefaultAutoOpen.test_erfa_found __________________________________________
self = <astropy.utils.iers.tests.test_leap_second.TestDefaultAutoOpen object at ...>

    def test_erfa_found(self):
        # Set huge maximum age such that whatever ERFA has is OK.
        # Since it is checked first, it should thus be found.
        with iers.conf.set_temp('auto_max_age', 100000):
            ls = iers.LeapSeconds.open()
>       assert ls.meta['data_url'] == 'erfa'
E       AssertionError: assert 'https://hpie...ap_Second.dat' == 'erfa'
E         - erfa
E         + https://hpiers.obspm.fr/iers/bul/bulc/Leap_Second.dat

astropy/utils/iers/tests/test_leap_second.py:243: AssertionError
________________________________________ TestDefaultAutoOpen.test_builtin_found ________________________________________
self = <astropy.utils.iers.tests.test_leap_second.TestDefaultAutoOpen object at ...>

    def test_builtin_found(self):
        # Set huge maximum age such that built-in file is always OK.
        # If we remove 'erfa', it should thus be found.
        self.remove_auto_open_files('erfa')
        with iers.conf.set_temp('auto_max_age', 100000):
            ls = iers.LeapSeconds.open()
>       assert ls.meta['data_url'] == iers.IERS_LEAP_SECOND_FILE
E       AssertionError: assert 'https://hpie...ap_Second.dat' == '/home/.../Do...ap_Second.dat'
E         - /home/.../astropy/utils/iers/data/Leap_Second.dat
E         + https://hpiers.obspm.fr/iers/bul/bulc/Leap_Second.dat

astropy/utils/iers/tests/test_leap_second.py:251: AssertionError
______________________________________ TestDefaultAutoOpen.test_fake_future_file _______________________________________
self = <astropy.utils.iers.tests.test_leap_second.TestDefaultAutoOpen object at ...>
tmpdir = local('/tmp/.../pytest-16/test_fake_future_file0')

    def test_fake_future_file(self, tmpdir):
        fake_file = make_fake_file('28 June 2345', tmpdir)
        # Try as system file for auto_open, setting auto_max_age such
        # that any ERFA or system files are guaranteed to be expired.
        with iers.conf.set_temp('auto_max_age', -100000), \
                iers.conf.set_temp('system_leap_second_file', fake_file):
            ls = iers.LeapSeconds.open()
>       assert ls.expires == Time('2345-06-28', scale='tai')
E       AssertionError: assert <Time object: scale='tai' format='iso' value=2022-06-28 00:00:00.000>
== <Time object: scale='tai' format='iso' value=2345-06-28 00:00:00.000>
E        +  where <Time object: scale='tai' format='iso' value=2022-06-28 00:00:00.000>
= <LeapSeconds length=42>\n
 year month tai_utc \n
int32 int32 float64 \n
----- ----- --------\n
 1960     1 1.417818\n 1961     ...    1     32.0\n
 2006     1     33.0\n
 2009     1     34.0\n
 2012     7     35.0\n 2015     7     36.0\n
 2017     1     37.0.expires
E        +  and   <Time object: scale='tai' format='iso' value=2345-06-28 00:00:00.000>
= Time('2345-06-28', scale='tai')

astropy/utils/iers/tests/test_leap_second.py:272: AssertionError
```

**Comment by pllim** [comment]
@mhvk , I don't grok why two tests fail with the change you suggested. Any advise on how to proceed? Thanks!

File: astropy/utils/iers/iers.py
Comment (mhvk) [inline astropy/utils/iers/iers.py:987]: Ah, a little mistake: this should be `TimeDelta(offset, format='jd')`, otherwise we do the `180-` twice!

**Comment by mhvk** [comment]
I think I found the mistake. If the first test still fails, it is fine to adjust it -- I think it is an improvement to use the best of the available leap second tables rather than just the first even for `auto_max_age = None`.

File: astropy/utils/iers/iers.py
Comment (pllim) [reply astropy/utils/iers/iers.py]: Good catch. Thanks!

**Comment by pllim** [comment]
Something still isn't quite right...

```
        # Use the fake files to make sure auto_max_age is safe.
        # Should have no warning here.
        with iers.conf.set_temp('auto_max_age', None):
            ls3 = iers.LeapSeconds.auto_open([fake_file1,
                                              iers.IERS_LEAP_SECOND_FILE])
>       assert ls3.meta['data_url'] == fake_file1
E       AssertionError: assert '/home/runner...ap_Second.dat' == '/tmp/pytest-...p_seconds.dat'
E         - /tmp/pytest-of-runner/pytest-0/test_fake_expired_file0/fake_leap_seconds.dat
E         + /home/runner/work/astropy/astropy/.tox/py310-test-cov/lib/python3.10/site-packages/astropy/utils/iers/data/Leap_Second.dat

../../.tox/py310-test-cov/lib/python3.10/site-packages/astropy/utils/iers/tests/test_leap_second.py:193: AssertionError
```

**Comment by pllim** [comment]
I don't understand this test... It expects the algorithm to actually pick the stale one, it seems like.

```python
        # Between these and the built-in one, the built-in file is best.
        ls = iers.LeapSeconds.auto_open([fake_file1, fake_file2,
                                         iers.IERS_LEAP_SECOND_FILE])
        assert ls.meta['data_url'] == iers.IERS_LEAP_SECOND_FILE

        # ...

        # Use the fake files to make sure auto_max_age is safe.
        # Should have no warning here.
        with iers.conf.set_temp('auto_max_age', None):
            ls3 = iers.LeapSeconds.auto_open([fake_file1,
                                              iers.IERS_LEAP_SECOND_FILE])
        assert ls3.meta['data_url'] == fake_file1
```

**Comment by mhvk** [comment]
I think that test has simple changed because the result is now better.  I've pushed the change (and added a bit to it for good measure).

**Review by mhvk — APPROVED** [review]
I think this is all OK now!

**Comment by pllim** [comment]
Thanks, @mhvk ! I'll wait a bit to see if this actually helps with @weaverba137 's use case before merge.

**Comment by weaverba137** [comment]
OK, while this *does* eliminate the unnecessary warnings, I am not convinced that @mhvk's approach is correct.  Remember, we are working in an HPC environment and must prevent any downloads.  We *do not* want the value of `good_enough` to revert to default when `auto_max_age = None`.  We want `good_enough` to have a value that will always prevent a download.

Can you walk me through the logic of this latest approach and demonstrate how this satisfies this requirement?

File: astropy/utils/iers/iers.py
Comment (weaverba137) [inline astropy/utils/iers/iers.py:986]: As noted in a previous comment, this has the affect of reverting `auto_max_age` to its default value when it is `None`.  I don't think this was the intended functionality.

**Review by weaverba137 — COMMENTED** [review]
As mentioned previously, I think this comes down to knowing exactly how `good_enough` works in this method.

**Comment by mhvk** [comment]
@weaverba137 - `conf.auto_download` ensures that no file is ever downloaded (I think, do correct me if I'm wrong!! see line 1004). This PR was really just to get rid of the warning that the file would be stale. But when I looked at the code, I realized that among files already in the cache or otherwise on the computer, the old code would just pick the first rather than the best, which seemed a mistake, so I corrected it to pick the first that clearly is good enough or the best available.

**Comment by weaverba137** [comment]
Aha, that makes more sense.  Let me take one more look at the full code in that light.

**Comment by weaverba137** [comment]
@mhvk, I am satisfied that the changes here will not inadvertently cause a download.  However, I don't see the change you made to pick the "best" file over the "first" file. Is that change in this PR?

**Review by weaverba137 — APPROVED** [review]
@mhvk, thank you for your clarifications.  I think this is ready to go.

File: astropy/utils/iers/iers.py
Comment (pllim) [reply astropy/utils/iers/iers.py:986]: For future reference: This point has been clarified in subsequent discussions in this PR.

**Review by pllim — COMMENTED** [review]
Thanks for the reviews! @mhvk , do you want me to rebase/squash? If not, feel free to merge.

**Comment by mhvk** [comment]
Let's just get it in!  Thanks, all! Slowly but surely we are making sure astropy rocks on every platform!

**Pull request merged**
